tests/integration/rosetta/rosetta_test.go

```diff
@@ -0,0 +1,253 @@
+					Address: "8c5303eaa26202d6",
+				},
+				BlockID: identifier.Block{
+					Index: 0,
+					Hash:  "d47b1bf7f37e192cf83d2bee3f6332b0d9b15c0aa7660d1e5322ea964667b333",
+				},
+				Currencies: getDefaultCurrencySpec(),
+			},
+			wantBalance:    "0",
+			wantStatusCode: http.StatusOK,
+			wantHandlerErr: assert.NoError,
+		},
+		{
+			name:           "empty balance request",
+			request:        rosetta.BalanceRequest{},
+			wantStatusCode: http.StatusUnprocessableEntity,
+			wantHandlerErr: assert.Error,
+		},
+		{
+			name: "block hash and height mismatch",
+			request: rosetta.BalanceRequest{
+				NetworkID: getDefaultNetworkID(),
+				AccountID: identifier.Account{
+					Address: "8c5303eaa26202d6",
+				},
+				BlockID: identifier.Block{
+					Index: 99,
+					Hash:  "d47b1bf7f37e192cf83d2bee3f6332b0d9b15c0aa7660d1e5322ea964667b333",
+				},
+				Currencies: getDefaultCurrencySpec(),
+			},
+			wantStatusCode: http.StatusUnprocessableEntity,
+			wantHandlerErr: assert.Error,
+		},
+		{
+			name: "invalid account address",
+			request: rosetta.BalanceRequest{
+				NetworkID: getDefaultNetworkID(),
+				AccountID: identifier.Account{
+					Address: "invalid_address",
+				},
+				BlockID: identifier.Block{
+					Index: 0,
+					Hash:  "d47b1bf7f37e192cf83d2bee3f6332b0d9b15c0aa7660d1e5322ea964667b333",
+				},
+				Currencies: getDefaultCurrencySpec(),
+			},
+			wantStatusCode: http.StatusUnprocessableEntity,
+			wantHandlerErr: assert.Error,
+		},
+	}
+
+	for _, test := range tests {
+
+		t.Run(test.name, func(t *testing.T) {
+
+			// prepare request payload (JSON)
+			enc, err := json.Marshal(test.request)
+			require.NoError(t, err)
+
+			// create request
+			req := httptest.NewRequest(http.MethodPost, "/account/balance", bytes.NewReader(enc))
+			req.Header.Set(echo.HeaderContentType, echo.MIMEApplicationJSON)
+
+			rec := httptest.NewRecorder()
+
+			ctx := echo.New().NewContext(req, rec)
+
+			// execute the request
+			err = rosettaSvc.Balance(ctx)
+			test.wantHandlerErr(t, err)
+
+			// validate response data
+
+			// validate status code
+			if test.wantStatusCode != http.StatusOK {
+
+				// ugly workaround for HTTP status code set by router vs set on context
+				e, ok := err.(*echo.HTTPError)
+				require.True(t, ok)
+				assert.Equal(t, test.wantStatusCode, e.Code)
+
+				// nothing more to do, response validation should only be done for '200 OK' responses
+				return
+			}
+
+			assert.Equal(t, test.wantStatusCode, rec.Result().StatusCode)
+
+			// unpack response
+			var balanceResponse rosetta.BalanceResponse
+			require.NoError(t, json.Unmarshal(rec.Body.Bytes(), &balanceResponse))
+
+			// verify that the block data matches the input data
+			assert.Equal(t, test.request.BlockID.Index, balanceResponse.BlockID.Index)
+			assert.Equal(t, test.request.BlockID.Hash, balanceResponse.BlockID.Hash)
+
+			// verify that we have at least one balance in the response
+			assert.Len(t, balanceResponse.Balances, 1)
```

Comment:
Don't we know the exact amount of balances there should be?

Reply:
My original intent was to stop the test there, since there's no point in checking the content of the first element of an array if it's empty (the checks below this one). It would still fail, but be more verbose than needed I think.

But the assert wouldn't do that, `require` would.

But after rethinking, I think it's better that I just do 
```go
if len(balanceResponse.Balances) > 0 {
    // check the value of balanceResponse.Balances[0]
}
```

That way if any new checks get added to the bottom of test, they wouldn't be skipped because of the response array check.

Reply:
Oh I thought that since we would know that we are supposed to have, for example, exactly 13 balances, we would write something along the lines of:

```go
if assert.Len(t, balanceResponse.Balances, 13) {
	// Check the value of balanceResponse.Balances[0]
}
```

It has the same effect as the current code but ensures that we have the exact amount we expect.

Reply:
Ah, nice, that's much cleaner. I'll use that :)